File: src/sbbs3/qtmonitor/login_attempts_widget.py
"""Failed login attempts display widget."""

from datetime import datetime

from PySide6.QtCore import Qt, Slot, Signal
from PySide6.QtGui import QColor
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QTreeWidget, QTreeWidgetItem, QMenu,
)


class LoginAttemptsWidget(QWidget):
    """Displays failed login attempts from MQTT updates."""

    count_changed = Signal(int)  # total number of IPs with failed attempts

    def __init__(self, dark=True, parent=None):
        super().__init__(parent)
        self._attempts = {}
        self._dark = dark
        self._setup_ui()

    def _setup_ui(self):
        layout = QVBoxLayout(self)
        layout.setContentsMargins(2, 2, 2, 2)

        self._tree = QTreeWidget()
        self._tree.setHeaderLabels([
            "IP Address", "First Attempt", "Last Attempt",
            "Count", "Dupes", "Protocol", "Username",
        ])
        self._tree.setColumnCount(7)
        self._tree.setRootIsDecorated(False)
        self._tree.setAlternatingRowColors(True)
        self._tree.setSortingEnabled(True)
        self.set_dark(self._dark)
        layout.addWidget(self._tree)

    def set_dark(self, dark):
        self._dark = dark
        if dark:
            self._tree.setStyleSheet(
                "QTreeWidget { background-color: #1e1e1e; color: #cccccc; "
                "alternate-background-color: #252525; }"
            )
        else:
            self._tree.setStyleSheet("")

    @Slot(str, str, dict)
    def update_attempt(self, ip, action, data):
        if action == "clear":
            if ip in self._attempts:
                idx = self._tree.indexOfTopLevelItem(self._attempts[ip])
                if idx >= 0:
                    self._tree.takeTopLevelItem(idx)
                del self._attempts[ip]
            self.count_changed.emit(len(self._attempts))
            return

        if ip not in self._attempts:
            item = QTreeWidgetItem()
            self._tree.addTopLevelItem(item)
            self._attempts[ip] = item

        item = self._attempts[ip]
        item.setText(0, ip)
        for col, key in [(1, "first"), (2, "last")]:
            ts = data.get(key, "")
            try:
                dt = datetime.fromisoformat(ts)
                item.setText(col, dt.strftime("%b %d %H:%M:%S"))
            except (ValueError, TypeError):
                item.setText(col, ts)
        item.setText(5, data.get("protocol", ""))
        item.setText(6, data.get("username", ""))

        try:
            count = int(data.get("count", "0"))
        except ValueError:
            count = 0
        item.setData(3, Qt.ItemDataRole.DisplayRole, count)

        try:
            dupes = int(data.get("dupes", "0"))
        except ValueError:
            dupes = 0
        item.setData(4, Qt.ItemDataRole.DisplayRole, dupes)
        if count >= 10:
            for col in range(7):
                item.setForeground(col, QColor("#FF4444") if self._dark else QColor("red"))
        elif count >= 5:
            for col in range(7):
                item.setForeground(col, QColor("#FF00FF") if self._dark else QColor("magenta"))
        self.count_changed.emit(len(self._attempts))

File: src/sbbs3/qtmonitor/loginattemptswidget.cpp
#include "loginattemptswidget.h"
#include "textutil.h"
#include <QVBoxLayout>
#include <QDateTime>
#include <QMenu>
#include <QHBoxLayout>

class SortableItem : public QTreeWidgetItem
{
public:
	bool operator<(const QTreeWidgetItem &other) const override {
		int col = treeWidget()->sortColumn();
		QVariant a = data(col, Qt::UserRole);
		QVariant b = other.data(col, Qt::UserRole);
		if (a.typeId() == QMetaType::QDateTime && b.typeId() == QMetaType::QDateTime)
			return a.toDateTime() < b.toDateTime();
		return QTreeWidgetItem::operator<(other);
	}
};

LoginAttemptsWidget::LoginAttemptsWidget(const QStringList &servers,
                                         const QHash<QString, QString> &serverLabels,
                                         bool dark, QWidget *parent)
	: QWidget(parent), m_dark(dark)
{
	auto *layout = new QVBoxLayout(this);
	layout->setContentsMargins(2, 2, 2, 2);

	auto *toolbar = new QHBoxLayout;
	auto *clearBtn = new QToolButton;
	clearBtn->setText("Clear");
	clearBtn->setPopupMode(QToolButton::InstantPopup);
	auto *clearMenu = new QMenu(this);
	for (const auto &server : servers) {
		auto *act = clearMenu->addAction(serverLabels.value(server, server));
		connect(act, &QAction::triggered, this, [this, server] {
			emit clearAllAttempts(server);
		});
	}
	clearMenu->addSeparator();
	auto *allAct = clearMenu->addAction("All");
	connect(allAct, &QAction::triggered, this, [this, servers] {
		for (const auto &server : servers)
			emit clearAllAttempts(server);
	});
	clearBtn->setMenu(clearMenu);
	toolbar->addWidget(clearBtn);
	toolbar->addStretch();
	layout->addLayout(toolbar);

	m_tree = new QTreeWidget;
	m_tree->setHeaderLabels({"IP Address", "First Attempt", "Last Attempt",
	                          "Count", "Dupes", "Protocol", "Username"});
	m_tree->setColumnCount(7);
	m_tree->setRootIsDecorated(false);
	m_tree->setAlternatingRowColors(true);
	m_tree->setSortingEnabled(true);
	m_tree->setSelectionMode(QAbstractItemView::ExtendedSelection);
	m_tree->setContextMenuPolicy(Qt::CustomContextMenu);
	connect(m_tree, &QTreeWidget::customContextMenuRequested, this, [this](const QPoint &pos) {
		auto selected = m_tree->selectedItems();
		if (selected.isEmpty())
			return;
		QMenu menu(this);
		auto *clearAct = menu.addAction(selected.size() == 1
			? "Clear " + selected.first()->text(0)
			: QString("Clear %1 entries").arg(selected.size()));
		connect(clearAct, &QAction::triggered, this, [this, selected] {
			for (auto *item : selected)
				emit clearAttempt(item->text(0));
		});
		menu.exec(m_tree->viewport()->mapToGlobal(pos));
	});
	setDark(m_dark);
	layout->addWidget(m_tree);
}

void LoginAttemptsWidget::updateAttempt(const QString &ip, const QString &action, const QVariantMap &data)
{
	if (action == "clear") {
		auto it = m_attempts.find(ip);
		if (it != m_attempts.end()) {
			delete m_tree->takeTopLevelItem(m_tree->indexOfTopLevelItem(*it));
			m_attempts.erase(it);
		}
		emit countChanged(m_attempts.size());
		return;
	}

	QTreeWidgetItem *item;
	auto it = m_attempts.find(ip);
	if (it == m_attempts.end()) {
		item = new SortableItem;
		m_tree->addTopLevelItem(item);
		m_attempts[ip] = item;
	} else {
		item = *it;
	}

	item->setText(0, ip);
	for (auto &[col, key] : std::initializer_list<std::pair<int, const char *>>{{1, "first"}, {2, "last"}}) {
		QString ts = data.value(key).toString();
		QDateTime dt = QDateTime::fromString(ts, Qt::ISODate);
		if (!dt.isValid())
			dt = QDateTime::fromString(ts.left(15), "yyyyMMdd'T'HHmmss");
		item->setText(col, dt.isValid() ? dt.toString("MMM dd hh:mm:ss") : ts);
		item->setData(col, Qt::UserRole, dt.isValid() ? dt : QDateTime());
	}
	setItemText(item, 5, data.value("protocol").toString());
	setItemText(item, 6, data.value("username").toString());

	int count = data.value("count", "0").toString().toInt();
	item->setData(3, Qt::DisplayRole, count);
	item->setData(4, Qt::DisplayRole, data.value("dupes", "0").toString().toInt());

	QColor color;
	if (count >= 10)
		color = m_dark ? QColor("#FF4444") : QColor("red");
	else if (count >= 5)
		color = m_dark ? QColor("#FF00FF") : QColor("magenta");
	if (color.isValid())
		for (int i = 0; i < 7; ++i)
			item->setForeground(i, color);

	emit countChanged(m_attempts.size());
}

void LoginAttemptsWidget::setDark(bool dark)
{
	m_dark = dark;
	m_tree->setStyleSheet(dark
		? "QTreeWidget { background-color: #1e1e1e; color: #cccccc; alternate-background-color: #252525; }"
		: QString());
}
